### DISCUSSION TITLE
Removing light polution

### DISCUSSION THREAD

**Discussion by effgarces** [AllskyTeam/allsky-modules#82] [General]
I'm trying to implement jbeale script (https://forums.raspberrypi.com/viewtopic.php?p=2096715#p2085037) for removing light polution in a module, but after trying multiple approaches I can never get it to apply. The sh script works perfectly, so I think that the problem lies in the module. For testing I simply duplicated the allsky_script.py to pass on the currentimage as parameter for the script. I'm hoping that someone can give an insight at what I'm doing wrong

[lightpollution.sh](https://gist.github.com/effgarces/02f42074c4a997ee2ec833ec867c0d19)
[allsky_lightpollution.py](https://gist.github.com/effgarces/d79e3941a5b6565810469de173658a9e)

**Comment by EricClaeys** [comment]
What's in /var/log/allsky.log?
If the script is failing there should be an error message.

**Comment by Alex-developer** [comment]
Unfortunately that approach will not work in a module. The current image is held in memory for faster processing by modules. 

The process goes like this

- Load image module loads the latest image into memory
- Each module can then modify the image in memory
- Save image finally writes the image to disk

In you module the image probably is being updated but then overwritten by the final save image module.

You could try re reading the image after your module has run the shell script, I have modified your code below. I have tried this and its very very slow. It takes around 25 seconds on my pi4 with an image taken indoors. It may be faster on an image taken outdoors

```
""" allsky_lightpollution.py

This module will modify the final image and apply jbeale script
"""
import allsky_shared as s
import os 
import subprocess
import cv2

metaData = {
    "name": "Remove light pollution",
    "description": "Removes light pollution",
    "events": [
        "day",
        "night"
    ],
    "module": "allsky_lightpollution"     
}

def lightpollution(params, event):
    script = "/home/pi/allsky/scripts/lightpollution.sh"

    if os.path.isfile(script):
        if os.access(script, os.X_OK):
            res = subprocess.check_output([script, s.CURRENTIMAGEPATH]) 
            s.image = cv2.imread(s.CURRENTIMAGEPATH)
            result = "Script {0} Executed.".format(script)
        else:
            s.log(0,"ERROR: Script {0} is not executable".format(script))
            result = "Script {0} Is NOT Executeable.".format(script)
    else:
        s.log(0,"ERROR: cannot access {0}".format(script))
        result = "Script {0} Not FOund.".format(script)

    return result`

**Reply by effgarces** [reply]
Thank you, I don't know why but I mistakenly assumed that the next step of process would re-read the file.
Yes this works perfectly, in my setup it does not take that much time, a couple of seconds at most, but that is probably because I'm using binning so as generate smaller files. 
I've uploaded the code https://github.com/effgarces/allsky_lightpollution, maybe it will be useful for others

**Reply by Alex-developer** [reply]
Glad you got it working.

If you are happy for me too I will create it as a module in here

**Reply by effgarces** [reply]
Yes sure, go ahead, hopefully it can be useful to someone .

**Reply by Alex-developer** [reply]
Thanks I'll add it to the list and close this dicussion

**Comment by Alex-developer** [comment]
> Yes sure, go ahead, hopefully it can be useful to someone .
